File: frontend/src/types/billing.ts
/** Billing TypeScript interfaces matching backend Pydantic schemas. */

export interface PlanResponse {
  id: string;
  name: string;
  description: string;
  tier: string;
  price_monthly: number;
  price_yearly: number;
  api_calls_limit: number;
  storage_gb_limit: number;
  agents_limit: number;
  features: string[];
}

export interface PlanScoredResponse {
  plan: PlanResponse;
  score: number;
  label: "good_fit" | "overpaying" | "at_risk" | "exceeds_limits";
}

export interface SubscriptionResponse {
  id: string;
  plan: PlanResponse;
  status: string;
  current_period_start: string | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
}

export interface CreateSubscriptionRequest {
  plan_id: string;
  billing_cycle: "monthly" | "yearly";
}

export interface ChangePlanRequest {
  new_plan_id: string;
}

export interface CancelSubscriptionRequest {
  immediate: boolean;
}

export interface UsageMeterResponse {
  metric_name: string;
  current: number;
  limit: number;
  percent_used: number;
}

export interface UsageForecastResponse {
  metric_name: string;
  current: number;
  projected_end_of_period: number;
  limit: number;
  percent_projected: number;
  exceeds_limit: boolean;
}

export interface InvoiceResponse {
  id: string;
  amount_usd: number;
  tax_usd: number;
  total_usd: number;
  status: string;
  issued_at: string | null;
  due_at: string | null;
  paid_at: string | null;
  pdf_url: string;
}

export interface InvoiceListResponse {
  items: InvoiceResponse[];
  total: number;
  page: number;
  page_size: number;
}

export interface RecommendationResponse {
  recommended_plan: PlanResponse;
  reasoning: string;
  savings_estimate_monthly: number;
  all_plans_scored: PlanScoredResponse[];
}

export interface CheckoutResponse {
  subscription_id: string | null;
  checkout_url: string | null;
  checkout_session_id?: string;
}

export interface CancelResponse {
  id: string;
  status: string;
  cancel_at_period_end: boolean;
}

export interface InvoicePdfResponse {
  invoice_id: string;
  pdf_url: string;
}

File: marketplace/schemas/billing.py
"""Pydantic schemas for billing endpoints: plans, subscriptions, usage, invoices."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Plans
# ---------------------------------------------------------------------------


class PlanFeatureItem(BaseModel):
    text: str
    included: bool = True


class PlanResponse(BaseModel):
    id: str
    name: str
    description: str = ""
    tier: str
    price_monthly: float
    price_yearly: float
    api_calls_limit: int
    storage_gb_limit: int
    agents_limit: int
    features: list[str] = Field(default_factory=list)


class PlanScoredResponse(BaseModel):
    """Plan with a fit score for recommendation."""

    plan: PlanResponse
    score: float
    label: str  # "good_fit" | "overpaying" | "at_risk" | "exceeds_limits"


# ---------------------------------------------------------------------------
# Subscriptions
# ---------------------------------------------------------------------------


class SubscriptionResponse(BaseModel):
    id: str
    plan: PlanResponse
    status: str
    current_period_start: datetime | None = None
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False


class CreateSubscriptionRequest(BaseModel):
    plan_id: str = Field(..., min_length=1, max_length=36)
    billing_cycle: Literal["monthly", "yearly"] = "monthly"


class ChangePlanRequest(BaseModel):
    new_plan_id: str = Field(..., min_length=1, max_length=36)


class CancelSubscriptionRequest(BaseModel):
    immediate: bool = False


# ---------------------------------------------------------------------------
# Usage
# ---------------------------------------------------------------------------


class UsageMeterResponse(BaseModel):
    metric_name: str
    current: float
    limit: int
    percent_used: float


class UsageForecastResponse(BaseModel):
    metric_name: str
    current: float
    projected_end_of_period: float
    limit: int
    percent_projected: float
    exceeds_limit: bool


# ---------------------------------------------------------------------------
# Invoices
# ---------------------------------------------------------------------------


class InvoiceResponse(BaseModel):
    id: str
    amount_usd: float
    tax_usd: float = 0.0
    total_usd: float
    status: str
    issued_at: datetime | None = None
    due_at: datetime | None = None
    paid_at: datetime | None = None
    pdf_url: str = ""


class InvoiceListResponse(BaseModel):
    items: list[InvoiceResponse]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# Recommendation
# ---------------------------------------------------------------------------


class RecommendationResponse(BaseModel):
    recommended_plan: PlanResponse
    reasoning: str
    savings_estimate_monthly: float
    all_plans_scored: list[PlanScoredResponse]
